### DISCUSSION TITLE
Best way to pass context props in server components?

### DISCUSSION THREAD

**Discussion by wit221** [vercel/next.js#45540] [Help] [answered]
After trying out various approaches to sharing contextual props (such as whether the request comes from mobile or desktop), including using react context, redux, as well as global variables, I am beginning to think my only option is to explicitly pass such props down the component hierarchy to every component that needs it, like so:

```
export const getServerSideProps: GetServerSideProps = async (props) => {
  const isMobileOnServer = isMobileServerSide(props.headers);

  return {
    props: {
      isMobileOnServer,
    },
  };
}
```

```
const MyPage: NextPage<MyPageProps> = (props) => {
  return <Layout isMobileOnServer={props.isMobileOnServer}/>
  // Pass that prop to only children of `Layout` that need it, and so on.
}
```

```
export const isMobileServerSide = (headers: IncomingHttpHeaders) => {
  const isMobileServer = Boolean(
    headers["user-agent"] &&
      /(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|ipad|iris|kindle|Android|Silk|lge |maemo|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino/i.test(
        headers["user-agent"]
      )
  );

  return isMobileServer;
};
```

This is bad because whether the request is mobile or not fits the react context approach much better (we don't want to pass props down the component tree like this). However, I failed to make this work with react context such that the component would render with the right value.

What is the recommended approach to the above issue? Is it possible to use react contexts on server-side components? If not, how can I achieve this?

**Comment by icyJoseph** [comment, marked as answer]
I understand that some business requirements, or just wanting to experiment are enough motivation, but rendering content for devices with this distinction is a bit strange. With that being said.

How did you fail to use React context in this way? you just pass the props into your provider, as value, and then read it downstream. 

Anyway, please do consider composition as well: https://www.youtube.com/watch?v=3XaXKiXtNjw

Could you please share code on what you tried to do with Context? Did you try to useEffect the initial state into the context?

Is it possible that you also have an `_app` file?

Please do share more info and I'll try to help you out. I don't want to make any assumptions just yet.
